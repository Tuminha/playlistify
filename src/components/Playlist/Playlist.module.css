.Playlist {
    width: 50%;
  }
  
  .Playlist input[type="text"] {
    width: 100%;
    padding: 10px;
    margin-bottom: 20px;
    font-size: 24px;
    font-weight: bold;
    color: #1DB954;
    border: none;
    outline: none;
  }
  
  .Playlist-save {
    padding: 15px 30px;
    cursor: pointer;
    background-color: #1DB954;
    color: white;
    border: none;
  }
  
  .Playlist-remove {
    padding: 10px;
    cursor: pointer;
    background-color: #FF0000;
    color: white;
    border: none;
  }

  
  