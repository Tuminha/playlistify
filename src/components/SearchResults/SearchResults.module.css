.SearchResults {
    width: 50%;
  }
  
  .SearchResults h2 {
    font-size: 24px;
    margin-bottom: 5px;
  }
  