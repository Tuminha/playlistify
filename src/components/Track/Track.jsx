import React from 'react';
import styles from './Track.module.css';


function Track({ track, onAdd, onRemove }) {
  return (
    <div className={styles.Track} key={track.id} data-testid="track">
      <div className={styles.TrackInformation}>
        <h3>{track.name}</h3>
        <p>{`${track.artist} | ${track.album}`}</p>
      </div>
      <button className={styles.TrackAction} onClick={() => onAdd(track)}>Add</button>
      {onRemove && <button className={styles.TrackAction} onClick={() => onRemove(track)}>Remove</button>}
    </div>
  );
}

export default Track;

