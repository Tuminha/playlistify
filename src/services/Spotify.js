let accessToken;
const clientId = '';
const clientSecret = '';
const redirectUri = 'http://localhost:3002/';

const Spotify = {
  getAccessToken() {
    console.log('Getting access token:', accessToken);
    if (accessToken) {
      return accessToken;
    }

    const accessTokenMatch = window.location.href.match(/access_token=([^&]*)/);
    const expiresInMatch = window.location.href.match(/expires_in=([^&]*)/);

    if (accessTokenMatch && expiresInMatch) {
      accessToken = accessTokenMatch[1];
      const expiresIn = Number(expiresInMatch[1]);
      window.setTimeout(() => accessToken = '', expiresIn * 1000);
      window.history.pushState('Access Token', null, '/');
      return accessToken;
    } else {
      const url = `https://accounts.spotify.com/authorize?client_id=${clientId}&response_type=token&scope=playlist-modify-public&redirect_uri=${redirectUri}`;
      if (process.env.NODE_ENV === 'test') {
        return url;
      } else {
        window.location = url;
      }
    }
  },
  setAccessToken(token) {
    console.log('Setting access token:', token);
    accessToken = token;
  },
  async search(term) {
    const accessToken = this.getAccessToken();
    let response;
    try {
      response = await fetch(`https://api.spotify.com/v1/search?type=track&q=${term}`, {
        headers: { Authorization: `Bearer ${accessToken}` }
      });
    } catch (error) {
      console.error(`Fetch error: ${error}`);
      return [];
    }
    if (!response) {
      console.error(`No response received.`);
      return [];
    }
    if (!response.ok) {
      console.error(`HTTP error! status: ${response.status}`);
      return [];
    } else {
      let jsonResponse;
      try {
        jsonResponse = await response.json();
      } catch (error) {
        console.error(`Response parsing error: ${error}`);
        return [];
      }
      if (!jsonResponse.tracks) {
        return [];
      }
      return jsonResponse.tracks.items.map(track => ({
        id: track.id,
        name: track.name,
        artist: track.artists[0].name,
        album: track.album.name,
        uri: track.uri
      }));
    }
  },
  // other methods to interact with Spotify API...
};

export default Spotify;

