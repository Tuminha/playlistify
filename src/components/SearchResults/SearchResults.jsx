import React from 'react';
import PropTypes from 'prop-types';
import Track from '../Track/Track';
import styles from './SearchResults.module.css';


function SearchResults({ tracks = [], onAdd }) {
  // Check if tracks prop is received and is an array
  if (!Array.isArray(tracks)) {
    console.error('Invalid prop: tracks should be an array');
    return null;
  }

  SearchResults.propTypes = {
    tracks: PropTypes.array,
    onAdd: PropTypes.func.isRequired,
  };

  // Render a Track component for each search result
  return (
    <div className={styles.SearchResults}>
      {tracks.map(track => (
        <Track key={track.id} track={track} onAdd={onAdd} />
      ))}
    </div>
  );
}



export default SearchResults;