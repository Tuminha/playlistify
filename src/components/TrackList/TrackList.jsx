import React from 'react';
import Track from '../Track/Track';
import styles from './TrackList.module.css';



function TrackList({ tracks = [], onRemove, onAdd }) {
  console.log("Rendering TrackList component");
  console.log("TrackList component received tracks: ", tracks); // Added console.log to understand the flow of data
  if (!tracks || tracks.length === 0) {
    return <div>No tracks found. Please try another search.</div>;
  } else {
    return (
      <div className={styles.TrackList} data-testid="track-list">
        {
          tracks.map(track => {
            console.log("Rendering track: ", track); // Added console.log to understand the flow of data
            return <Track key={track.id} track={track} onRemove={onRemove} onAdd={onAdd}/>
          })
        }
      </div>
    );
  }
}

export default TrackList;
