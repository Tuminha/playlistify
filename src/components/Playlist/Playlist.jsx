import React, { useState } from 'react';
import TrackList from '../TrackList/TrackList';
import './Playlist.module.css';

function Playlist({ playlistName = '', playlistTracks, onRemove, onNameChange, onAdd }) {
  console.log("Rendering Playlist component");
  
  // Use the playlistTracks prop to pass down to the TrackList component
  const tracks = playlistTracks;
  console.log("Tracks in Playlist: ", tracks);

  // Check if the track's name is being rendered as 'Track 1'
  tracks.forEach(track => {
    if (track.name !== 'Track 1') {
      console.error(`Track name is not being rendered as 'Track 1': ${track.name}`);
    }
  });

  // Use useState to manage the playlist name in the local state
  const [currentPlaylistName, setCurrentPlaylistName] = useState(playlistName);

  // Handle the change of the playlist name
  const handleNameChange = (event) => {
    setCurrentPlaylistName(event.target.value);
    onNameChange(event.target.value);
  };

  // Create an array of URIs for each track in the playlist
  const trackURIs = tracks.map(track => track.uri);

  // Function to save these URIs to the user's Spotify account
  const savePlaylist = () => {
    console.log("Saving playlist to Spotify: ", trackURIs);
    // Reset the playlist in the app
    setCurrentPlaylistName('');
    onNameChange('');
    onRemove(playlistTracks);
  };

  return (
    <div className="Playlist" data-testid="playlist">
      <input 
        value={currentPlaylistName} 
        className="Playlist-input" 
        placeholder="New Playlist" // Add placeholder text to the input field
        onChange={handleNameChange} // Handle the change of the playlist name
      />
      <TrackList tracks={tracks} onRemove={onRemove} onAdd={onAdd} className="TrackList" data-testid="track-list" />
      <button className="Playlist-save" onClick={savePlaylist}>SAVE TO SPOTIFY</button>
    </div>
  );
}

export default Playlist;