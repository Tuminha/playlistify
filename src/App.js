import React, { useState } from 'react';
import SearchBar from './components/SearchBar/SearchBar';
import SearchResults from './components/SearchResults/SearchResults';
import Playlist from './components/Playlist/Playlist';
import playlistStyles from './components/Playlist/Playlist.module.css';
import styles from './App.module.css';

function App() {
  const [searchResults, setSearchResults] = useState([]);
  const [tracks, setTracks] = useState([
    { id: 1, name: "Track 1", artist: "Artist 1", album: "Album 1" },
    { id: 2, name: "Track 2", artist: "Artist 2", album: "Album 2" },
    // ... more tracks
  ]);
  const [playlistName, setPlaylistName] = useState('New Playlist');
  const [playlistTracks, setPlaylistTracks] = useState([]);

  // Function to update playlist name
  const updatePlaylistName = (name) => {
    setPlaylistName(name);
  };

  // Function to add track to playlist
  const addToPlaylist = (track) => {
    if (!playlistTracks.find(savedTrack => savedTrack.id === track.id)) {
      setPlaylistTracks(prevTracks => [...prevTracks, track]);
      removeFromPlaylist(track);
    }
  };

  const removeFromPlaylist = (track) => {
    setPlaylistTracks(prevTracks => prevTracks.filter(t => t.id !== track.id));
  };

  const handleSearchResults = (results) => {
    setSearchResults(results);
  };

  return (
    <div className={styles.App} data-testid="app">
      <SearchBar onSearch={setSearchResults} />
      <div className={playlistStyles.AppPlaylist}>
        <SearchResults tracks={searchResults} onAdd={addToPlaylist} />
        <Playlist playlistName={playlistName} playlistTracks={playlistTracks} onNameChange={updatePlaylistName} onRemove={removeFromPlaylist} />
      </div>
    </div>
  );
}

export default App;

