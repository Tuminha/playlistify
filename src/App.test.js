import { render, screen, fireEvent, waitFor, within } from '@testing-library/react';
import App from './App';
import SearchResults from './components/SearchResults/SearchResults';
import Playlist from './components/Playlist/Playlist';
import Track from './components/Track/Track';
import TrackList from './components/TrackList/TrackList';
import Spotify from './services/Spotify';



const mockTracks = [
  { id: 1, name: "Track 1", artist: "Artist 1", album: "Album 1" },
  // Add more track objects here
];

global.fetch = jest.fn().mockImplementation(() => {
  return Promise.resolve({
    json: () => Promise.resolve({
      tracks: {
        items: [
          {
            id: '1',
            name: 'Track 1',
            artists: [{ name: 'Artist 1' }],
            album: { name: 'Album 1' },
            uri: 'uri1'
          },
          // Add more track objects here
          {
            id: '2',
            name: 'Track 2',
            artists: [{ name: 'Artist 2' }],
            album: { name: 'Album 2' },
            uri: 'uri2'
          },
          {
            id: '3',
            name: 'Track 3',
            artists: [{ name: 'Artist 3' }],
            album: { name: 'Album 3' },
            uri: 'uri3'
          }
        ]
      }
    })
  });
});




/* Test to verify that the SearchResults component renders a list of Track components:

Render the SearchResults with a mock array of tracks.
Check that for each track object, a Track component is rendered.
Test to ensure that each Track component receives the correct props:

Render the TrackList with a mock array of tracks.
For each Track component, check that it has received the correct track data as props.
Test to confirm the unique key for each Track component:

Render the SearchResults component.
Verify that each Track component has a unique key prop. */

describe('SearchResults', () => {
  test('renders a list of Track components', () => {
    render(<SearchResults tracks={mockTracks} />);
    const trackElements = screen.getAllByTestId('track');
    expect(trackElements.length).toBe(mockTracks.length);
  });

  test('each Track component receives the correct props', () => {
    render(<TrackList tracks={mockTracks} />);
    const trackElements = screen.getAllByTestId('track');
    trackElements.forEach((trackElement, index) => {
      expect(trackElement).toHaveTextContent(mockTracks[index].name);
      expect(trackElement).toHaveTextContent(mockTracks[index].artist);
      expect(trackElement).toHaveTextContent(mockTracks[index].album);
    });
  });

  test('each Track component has a unique key prop', () => {
    render(<SearchResults tracks={mockTracks} />);
    const trackElements = screen.getAllByTestId('track');
    const keyValues = trackElements.map(trackElement => trackElement.getAttribute('key'));
    const keySet = new Set(keyValues);
    expect(keyValues.length).toBe(keySet.size);
  });
} );

test('renders Playlist component with correct props', () => {
  render(<App />);
  const playlistElement = screen.getByTestId('playlist');
  expect(playlistElement).toBeInTheDocument();
  // Add more assertions here
});

// App.test.js
test('renders TrackList component with correct props', async () => {
  render(<Playlist playlistName="My Playlist" playlistTracks={ mockTracks } />);
  const trackListElement = await screen.findByTestId('track-list');
  expect(trackListElement).toBeInTheDocument();
  // Add more assertions here
});

test('renders a Track component for each track', () => {
  render(<TrackList tracks={ mockTracks } />);
  const trackElements = screen.getAllByTestId('track');
  expect(trackElements.length).toBe(mockTracks.length);
});







test('adds a track to the playlist when the "Add" button is clicked', async () => {
  render(<App />);

  // Simulate a search
  const mockOnAdd = jest.fn();
  const mockOnRemove = jest.fn();
  const mockTrack = { id: 1, name: "Track 1", artist: "Artist 1", album: "Album 1", uri: "spotify:track:1" };
  

  // Now the 'Add' buttons should be in the document
  render(<Track track={mockTrack} onAdd={mockOnAdd} onRemove={mockOnRemove} />);

  fireEvent.click(screen.getByText('Add'));

  expect(mockOnAdd).toHaveBeenCalledWith(mockTrack);



});


test('removes a track from the playlist when the "Remove" button is clicked', async () => {
  render(<App />);
  
  // Simulate a search
  const mockOnAdd = jest.fn();
  const mockOnRemove = jest.fn();
  const mockTrack = { id: 1, name: "Track 1", artist: "Artist 1", album: "Album 1", uri: "spotify:track:1" };
  
  // Now the 'Remove' buttons should be in the document
  render(<Track track={mockTrack} onAdd={mockOnAdd} onRemove={mockOnRemove} />);

  fireEvent.click(screen.getByText('Remove'));

  expect(mockOnRemove).toHaveBeenCalledWith(mockTrack);

});


test('allows the user to change the playlist name', () => {
  const mockOnNameChange = jest.fn();
  render(<Playlist playlistTracks={[]} onNameChange={mockOnNameChange} />);

  // Find the input field and type a new name into it
  const nameInput = screen.getByRole('textbox');
  fireEvent.change(nameInput, { target: { value: 'New Playlist Name' } });

  // Check if the input field now has the new name
  expect(nameInput.value).toBe('New Playlist Name');

  // Check if the onNameChange function was called with the new name
  expect(mockOnNameChange).toHaveBeenCalledWith('New Playlist Name');
});

test('saves the playlist to the user\'s Spotify account', () => {
  const mockTracks = [
    { id: 1, name: "Track 1", artist: "Artist 1", album: "Album 1", uri: "spotify:track:1" },
    // Add more track objects here
  ];
  const mockOnNameChange = jest.fn();
  const mockOnRemove = jest.fn();

  render(<Playlist playlistTracks={mockTracks} onNameChange={mockOnNameChange} onRemove={mockOnRemove} />);

  // Click the "SAVE TO SPOTIFY" button
  const saveButton = screen.getByText('SAVE TO SPOTIFY');
  fireEvent.click(saveButton);

  // Check if the onNameChange function was called with an empty string
  expect(mockOnNameChange).toHaveBeenCalledWith('');

  // Check if the onRemove function was called with the playlist tracks
  expect(mockOnRemove).toHaveBeenCalledWith(mockTracks);
});




describe('Spotify service', () => {
  beforeEach(() => {
    // Reset the access token before each test
    Spotify.setAccessToken(null);
  });

  it('should set and get access token correctly', () => {
    const testToken = 'test-token';
    Spotify.setAccessToken(testToken);
    expect(Spotify.getAccessToken()).toBe(testToken);
  });

  it('should return search results', async () => {
    // Log the fetch mock to see if it's being called and what it's returning
    console.log('fetch mock:', global.fetch);
  
    const results = await Spotify.search('test');
  
    // Log the results to see what's being returned
    console.log('results:', results);
  
    expect(results).toBeDefined();
    expect(results.length).toBeGreaterThan(0);
    // Add more assertions based on your mock track data
  });
});