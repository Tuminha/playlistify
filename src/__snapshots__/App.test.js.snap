// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visual Regression Testing renders correctly 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="SearchBar"
    >
      <input
        placeholder="Enter A Song, Album, or Artist"
      />
      <button
        class="SearchButton"
      >
        SEARCH
      </button>
    </div>
    <div
      class="AppPlaylist"
    >
      Loading...
      <div
        class="Playlist"
      >
        <input
          class="Playlist-input"
          value="New Playlist"
        />
        <p>
          No tracks in the playlist
        </p>
        <button
          class="Playlist-save"
        >
          SAVE TO SPOTIFY
        </button>
      </div>
    </div>
  </div>
</div>
`;
